### ISSUE TITLE
Intermittent rust test failure

### ISSUE THREAD

**Issue by petemoore** [taskcluster/taskcluster#6280]
**Describe the bug**
I had a rust test failure today that I've not seen before. My guess is that it is intermittent, and unrelated to my changes.

Issue was here: https://github.com/taskcluster/taskcluster/runs/13918398445

Looking at the code, it appears to write DATA to a file ('HELLO/WORLD') and then reads it from a file, checking that the content matches what was written to the file. It attempts to read the data twice. The first read was the one that failed in this case, so wasn't related to data being read twice.

My _guess_ is that perhaps:

```rs
file.write_all(DATA).await?;
```

assumes that the statement blocks until the data is flushed to the file and the file handle is closed. But perhaps code continues before the OS has flushed the data, so that the subsequent lines that read the data from the file don't see the new data that has been written to the file. This would explain why the left side is empty (no bytes) and the right side is the full data (the data that was meant to be written to the file) and also why the issue would be intermittent, because probably normally the OS immediately flushes the data:

```rs
assert_eq!(&copy_from_factory(&mut factory).await?, DATA);
```

@djmitche What do you think?

**Comment by djmitche** [comment]
I don't think this is a race -- operating systems are pretty good at indicating when data has been written to a file!

Even if the `.await` was missing on `file.write_all(DATA)`, I wouldn't expect intermittent behavior:
 * Rust warns for unused Futures
 * Futures that are dropped (which is what would happen without `.await`) don't do anything, so the result would be that the data is _never_ written to the file.

**Comment by djmitche** [comment]
Hm, this might be a tokio bug, though -- we actually duplicate the File (which we have to -- we only get one copy from `tempfile`) and perhaps that doesn't duplicate the buffer.

But [internally](https://docs.rs/tokio/1.26.0/src/tokio/fs/file.rs.html#93-96) it's using a `std::fs::File` directly, which just maps to OS calls -- no in-process buffering.

That said, adding a `file.flush().await?` does seem to stop this from reproducing. It's possible that updating `tokio` would fix some subtle bug here, too, although I don't see the bug.

**Comment by djmitche** [comment]
Oh! Tokio is running fs operations like `write` on a thread! But, it doesn't wait for those operations to complete when cloning.. That's a bug :)
